RavenDB-20265 - Optimizing Corax disk usage

ayende opened the pull request:
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-20265 

### Additional description

* Prevent degenerate case where Corax would consume more disk space than the source data because of limits in big posting lists.
* Moved small posting lists to use PFor encoding
* Use VxSort for speeding up large indexing batches
* More tests
* Fixed out of band access using AVX in PForDecoder (brute force manner)


### Type of change

- Bug fix

### How risky is the change?

- High

### Backward compatibility

- Breaking change

ayende committed:
RavenDB-20265 - Including large posting lists in the storage report calculations

ayende committed:
RavenDB-20265 - Fixing Container's free list handling so we'll be able to reuse pages more effectively

ayende committed:
RavenDB-20265 - Changing PFor encoding to natively handle 64 bits numbers, so we won't have to deal with fragmentation issues around entries that are very far apart from one another.

ayende committed:
RavenDB-20265 - Updating the Posting List to avoid the need to encoding entries that are far part in different pages.
We no longer need the Baseline for posting list leaf pages, since we can encoding 64 bits range directly in them.

ayende committed:
RavenDB-20265 - Changing the way we handle small posting lists to take advantage of PFor Encoding as well, now that they can take the full 64 bits range.

ayende committed:
RavenDB-20265 - Removed optimized code for growing the posting list downward, that isn't a scenario we care about, and the code looked suspect there.

ayende committed:
RavenDB-20265 Allow to encode 0 in PForEncoder

ayende committed:
RavenDB-20265
- Ensure that we cannot have branches that have just one leaf
- Use VxSort to optimize indexing
- Better tests for weird edge cases

ayende committed:
RavenDB-20265
- Allow to access NativeIntegersList directly to avoid Span overhead and sort using VxSort
- PForEncoder now uses raw pointers instead of Span
- Adding Verify() method to ensure the structure of the Posting List
- Avoid span access in PostingList.PrepareForCommit in favor of raw pointers
- Fixed an issue where we would always scan removals from the start of the page
- Adding more validations to the test

ayende committed:
RavenDB-20265 - Fixing polarity of check

ayende committed:
RavenDB-20265 - Simplifying the buffer management and ensuring that we always produce it in a consistent manner

ayende committed:
RavenDB-20265 - Ensure that we can't overflow the buffer by writing 0 bits to the byte _after_ the buffer

ayende committed:
RavenDB-20265 Ensure that PForDecoder will never use AVX to read beyond the size of the buffer. This is a blunt check to avoid the problem at the moment and still needs to be reviewed further.

ayende committed:
RavenDB-20265 Changing the Corax schema version

ayende committed:
RavenDB-20265 Fixing reading entries from small posting list to delete them

ayende committed:
RavenDB-20265 Support ListWithEmpty in facets

ayende committed:
RavenDB-20265 Actually support ListWithEmpty in facets

ayende committed:
RavenDB-20265 Rough cut for sorting optimization

redknightlois committed:
RavenDB-20265: Fill now also supports reentrancy and general Fill requirements.

ayende commented on src/Voron/Data/PostingLists/PForDecoder.cs:
```diff
@@ -156,9 +174,15 @@ private static int HorizontalOr(in Vector128<int> x)
+            if (maxByteOffset >= inputBufferSize)
+                throw new ArgumentOutOfRangeException();
+#endif
+            //TODO: Ensure that we aren't reading beyond the end of the buffer!
```
@redknightlois Note that this need a proper fix.

arekpalinski replied:
https://issues.hibernatingrhinos.com/issue/RavenDB-20295

redknightlois committed:
RavenDB-20265: Performance optimizations for SortingMatch

arekpalinski commented:
The test builds have been aborted 2 times already when whey were run on Linux. It looks that each time we got it when running test from `PostingListTestsExtended`:

1. https://jenkins.hibernatingrhinos.com/job/PR_Tests/5865/consoleFull

```
08:25:22    Passed SlowTests.Voron.PostingListTestsExtended.CanDeleteAndInsertInRandomOrder(seed: 1064156071, size: 796) [154 ms]
08:25:22    Passed SlowTests.Voron.PostingListTestsExtended.CanDeleteAndInsertInRandomOrder(seed: 506431817, size: 2) [151 ms]
08:25:22    Passed SlowTests.Voron.PostingListTestsExtended.CanDeleteAndInsertInRandomOrder(seed: 1047269048, size: 49853) [664 ms]
08:25:24    Passed SlowTests.Voron.PostingListTestsExtended.CanDeleteAndInsertInRandomOrder(seed: 1337, size: 200000) [2 s]
08:25:26    Passed SlowTests.Voron.PostingListTestsExtended.CanDeleteAndInsertInRandomOrder(seed: 511767612, size: 4172) [181 ms]
08:29:48  Cannot contact eagle2-ubuntu: hudson.remoting.ChannelClosedException: Channel "hudson.remoting.Channel@4f379617:JNLP4-connect connection from 10.0.0.116/10.0.0.116:59676": Remote call on JNLP4-connect connection from 10.0.0.116/10.0.0.116:59676 failed. The channel is closing down or has closed down
08:38:40  wrapper script does not seem to be touching the log file in /home/eagle-2/Jenkins/workspace/PR_Tests/s/test/SlowTests@tmp/durable-ac956de7
08:38:40  (JENKINS-48300: if on an extremely laggy filesystem, consider -Dorg.jenkinsci.plugins.durabletask.BourneShellScript.HEARTBEAT_CHECK_INTERVAL=86400)
08:38:40  [Pipeline] echo
08:38:40  [LOG] encountered hudson.AbortException: script returned exit code -1
```

2. https://jenkins.hibernatingrhinos.com/job/PR_Tests/5858/consoleFull

```
15:38:10    Passed SlowTests.Voron.PostingListTestsExtended.CanDeleteAndInsertInRandomOrder(seed: 1410139035, size: 17797681) [4 m 46 s]
15:38:10    Passed SlowTests.Voron.PostingListTestsExtended.CanDeleteAndInsertInRandomOrder(seed: 511767612, size: 4172) [179 ms]
15:41:47  Cannot contact eagle2-ubuntu: hudson.remoting.ChannelClosedException: Channel "hudson.remoting.Channel@286a0755:JNLP4-connect connection from 10.0.0.116/10.0.0.116:45806": Remote call on JNLP4-connect connection from 10.0.0.116/10.0.0.116:45806 failed. The channel is closing down or has closed down
15:50:24  wrapper script does not seem to be touching the log file in /home/eagle-2/Jenkins/workspace/PR_Tests/s/test/SlowTests@tmp/durable-9ea311d9
15:50:24  (JENKINS-48300: if on an extremely laggy filesystem, consider -Dorg.jenkinsci.plugins.durabletask.BourneShellScript.HEARTBEAT_CHECK_INTERVAL=86400)
15:50:24  [Pipeline] echo
15:50:24  [LOG] encountered hudson.AbortException: script returned exit code -1
```

arekpalinski commented:
On Windows we got the following failing tests:
https://jenkins.hibernatingrhinos.com/job/PR_Tests/5854/testReport/

More builds scheduled on Windows:
https://jenkins.hibernatingrhinos.com/job/PR_Tests/5866
https://jenkins.hibernatingrhinos.com/job/PR_Tests/5870

UPDATE: The following two are failing consistently on Windows:
- SlowTests.MailingList.Chad.CanQuerySpatialData(options: DatabaseMode = Single , SearchEngineMode = Corax)
- SlowTests.Tests.NoNonDisposableTests.ShouldExist

arekpalinski commented:
More builds scheduled on Linux: 
https://jenkins.hibernatingrhinos.com/job/PR_Tests/5869
https://jenkins.hibernatingrhinos.com/job/PR_Tests/5871

UPDATE: The above build also resulted in ` Cannot contact eagleX-ubuntu: ...` errors as well so the builds were aborted. 

We checked the machines logs. In both cases it failed due to `oom-killer`:

eagle-2

```
Apr 12 09:29:36 eagle2-Z390-GAMING-X kernel: Out of memory: Killed process 1437838 (dotnet) total-vm:284942748kB, anon-rss:14674308kB, file-rss:0kB, shmem-rss:146844kB, UID:1000 pgtables:38464kB oom_score_adj>
Apr 12 09:29:36 eagle2-Z390-GAMING-X systemd[1]: jenkins-agent-eagle2.service: A process of this unit has been killed by the OOM killer.
Apr 12 09:29:37 eagle2-Z390-GAMING-X systemd[1]: jenkins-agent-eagle2.service: Failed with result 'oom-kill'.
```

eagle-4

```
Apr 12 10:11:10 eagle-4 kernel: .NET Long Runni invoked oom-killer: gfp_mask=0x140dca(GFP_HIGHUSER_MOVABLE|__GFP_COMP|__GFP_ZERO), order=0, oom_score_adj=0
Apr 12 10:11:10 eagle-4 kernel: CPU: 12 PID: 877742 Comm: .NET Long Runni Tainted: G        W         5.19.0-35-generic #36~22.04.1-Ubuntu
Apr 12 10:11:10 eagle-4 kernel: Hardware name: Gigabyte Technology Co., Ltd. X570 GAMING X/X570 GAMING X, BIOS F3 06/13/2019
Apr 12 10:11:10 eagle-4 kernel: Call Trace:
Apr 12 10:11:10 eagle-4 kernel:  <TASK>
Apr 12 10:11:10 eagle-4 kernel:  show_stack+0x52/0x69
Apr 12 10:11:10 eagle-4 kernel:  dump_stack_lvl+0x49/0x6d
Apr 12 10:11:10 eagle-4 kernel:  dump_stack+0x10/0x18
Apr 12 10:11:10 eagle-4 kernel:  dump_header+0x53/0x248
Apr 12 10:11:10 eagle-4 kernel:  oom_kill_process.cold+0xb/0x10
Apr 12 10:11:10 eagle-4 kernel:  out_of_memory+0x106/0x300
Apr 12 10:11:10 eagle-4 kernel:  __alloc_pages_may_oom+0x117/0x1e0
Apr 12 10:11:10 eagle-4 kernel:  __alloc_pages_slowpath.constprop.0+0x459/0x810
```

maciejaszyk committed:
RavenDB-19603 Adjust boosting tests to BM25.

redknightlois committed:
RavenDB-20265: Microoptimization of the SortingMatch implementation.

redknightlois commented:
> UPDATE: The above build also resulted in ` Cannot contact eagleX-ubuntu: ...` errors as well so the builds were aborted.

Could this be related to https://github.com/ravendb/ravendb/pull/16243 ?

ayende commented on src/Corax/Queries/SortingMatch/SortingMatch.cs:
```diff
@@ -251,9 +257,13 @@ private struct Fetcher : IFetcher
                 for (int i = 0; i < read; i++)
                 {
-                    var cur = new MatchComparer<TComparer, TOut>.Item { Key = matches[i], };
+                    cur.Key = matches[i];
```
matchesRef here, to? To avoid the check?

redknightlois replied:
Yes. notices also that I am not doing a `new` anymore. I am using the same copy as an auxiliary to avoid the construction and initialization of memory for the struct. That is only visible at the assembler level.

ayende commented on src/Corax/Queries/SortingMatch/SortingMatch.cs:
```diff
@@ -287,12 +299,15 @@ private struct Fetcher : IFetcher
-                matches[i] = items[0].Key;
-                HeapDown(items, match._bufferUsedCount - i - 1);
+                ref var matchIdxPtr = ref Unsafe.Add(ref matchesRef, i);
+                matchIdxPtr = items[0].Key;
```
itemsStart  here, no?

redknightlois replied:
Yeah, we could use that. But in this case the check is done outside of the loop (loop hoisting). So no need to uglify that much.

arekpalinski commented:
> Could this be related to https://github.com/ravendb/ravendb/pull/16243 ?

You can try to rebase (and solve the conflict) to check it but I don't think it's related. I think we introduced a new memory issue here which manifests only on Linux for some reason.

ayende committed:
RavenDB-20265 - Fixing merge issue with accounting for deleted entries, removing unused method

ayende commented:
Pushed with rebase on top of current.
`CanUseFilterQueryOnMapIndexes` known to fail, see: https://issues.hibernatingrhinos.com/issue/RavenDB-20294/CanUseFilterQueryOnMapIndexes-fails-on-Corax

Will not investigate the other causes

ayende commented:
Testing the `test.CanDeleteAndInsertInRandomOrder(seed: 511767612, size: 4172);` under Linux, no error.

ayende commented:
Tested:

```
 test.CanDeleteAndInsertInRandomOrder(seed: 1064156071, size: 796);
 test.CanDeleteAndInsertInRandomOrder(seed: 506431817, size: 2);
 test.CanDeleteAndInsertInRandomOrder(seed: 1047269048, size: 49853);
 test.CanDeleteAndInsertInRandomOrder(seed: 1337, size: 200000);
 test.CanDeleteAndInsertInRandomOrder(seed: 511767612, size: 4172);
```

ayende committed:
RavenDB-18461 - Corax is not supporting compaction

ayende committed:
RavenDB-20265 - Marking test as no disposal needed

ayende committed:
RavenDB-20265 - Fixing AndWith should take into account only valid entries in the buffer

ayende committed:
RavenDB-20294 Refactor CoraxIndexReadOperation.Query to reduce nesting of operations

ayende committed:
RavenDB-20294 CanUseFilterQueryOnMapIndexes fails on Corax

ayende committed:
RavenDB-20294
- Do not introduce a new field for the index inside the test
- Tabs => Spaces

arekpalinski commented on test/Tests.Infrastructure/NonMacFactAttribute.cs:
```diff
@@ -0,0 +1,15 @@
+
+namespace FastTests;
+
+public class NonMacFactAttribute : FactAttribute
```
We already have `RavenMultiplatformFactAttribute` for that

ayende committed:
RavenDB-20294
- Ensuring that we handle paging + skipping properly on Corax
- Ensuring that the entry id will be the tie breaker in sorting in the case of identical values that we are sorting on

ayende committed:
RavenDB-20294 - Removed page size limit inside of Corax, we are already doing the limits at the request level

ayende committed:
RavenDB-20294 - Moving very slow test to StressTests, ensuring that ElectricFence test will run on Linux / Windows only

maciejaszyk committed:
RavenDB-20294 Corax sorting: when field is different type (or it's `NULL`) we've to set it as failed to read to get proper order. (since `.Read(...)`) can return wrong value.

maciejaszyk committed:
RavenDB-20265: In the case of sorting, we have to fetch all the data from storage to create proper paging. Sorting partial data may lead to incorrect overall results.
